docs/src/intro/usage.md

```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
 ```
 ./target/release/miden run -a miden/examples/fib/fib.masm -n 1
 ```
-This will run the example code to completion and will output the top element remaining on the stack.
\ No newline at end of file
+This will run the example code to completion and will output the top element remaining on the stack.
+
+## REPL
+
+The Miden Read–eval–print loop or repl for short is a Miden shell that allows for quick and easy debugging with Miden assembly. When the repl gets initialised, there are a bunch of awesome things an user can do like execute any Miden instruction, undo executed instructions, check the state of the stack and memory at anytime and more! A `history.txt` file is saved when the repl is exited. One thing to note that all the repl native commands starts with an `!` to differentiate them with regular assembly instruction. The Repl currently supports the following commands:
+
+### Miden assembly instruction
+
+All the Miden instruction mentioned in the Miden Assembly document is valid. (Ex. push.0, drop, dropw, swap, cswap, u32checked_add, mem_loadw, mtree_get etc.). One can either input instructions one by one or multiple instructions in one input.
+
+For example the below two commands will result in the same output.
+```
+>> push.1
+>> push.2
+>> push.3
+```
+
+```
+push.1 push.2 push.3
+```
+
+In order to execute a control flow operation, one needs to write the entire flow operation in a single line with spaces between individual operations.
+
+```
+repeat.20
+    pow2
+end
+```
+
+The above example should be written as following in the Repl tool:
+
+```
+repeat.20 pow2 end
+```
+
+### !help
+
+`!help` command prints out the list of all the available commands in the repl tool. 
+
+### !program
+
+`!program` command prints out the entire miden program being executed so far. For e.g. in the below scenario:
+
+```
+>> push.1.2.3.4
+>> repeat.16 pow2 end
+>> u32checked_add
+
+>> !program
+begin
+    push.1.2.3.4
+    repeat.16 pow2 end
+    u32checked_add
+end
+```
+
+### !stack
+
+`!stack` command prints out the state of the stack at the latest clock cycle. The state also contains the overflow part of the stack. The stack will be empty at initialisation i.e. it will only have 16 `0`. 
```

Comment:
Change to:

`!stack` command prints out the state of the stack at the last executed instruction. Since the stack always contains at least 16 elements, 16 or more elements will be printed out (even if all of them are zeros).